OAK-9447: Upgrade Mongo java driver to 5.2

Joscorbe opened the pull request:
Starting from base PR created by [rgambelli](https://github.com/rgambelli) #1761 

Some conflicts still need to be resolved before merging

Joscorbe committed:
OAK-9447: Upgrade Mongo java driver to 5.2 (#1787)

* OAK-9447

* Conflict with trunk resolved

* Conflict resolved

* conflict resolved

* Reordered imports like original classes and fixed oak-run about new
mongodb-driver-sync

* Reordered imports like original classes

* Upgraded com.mongodb and org.bson versions in Import-Package

* Restored original formatting

* Restored original import ordering

* Restored original import ordering

---------

Co-authored-by: raffaega <raffaega@CI00298583>
Co-authored-by: raffaega <raffaega@host.docker.internal>
Co-authored-by: Raffaele Gambelli <r.gambelli@gmail.com>
Co-authored-by: Raffaele Gambelli <raffaele.gambelli@cegeka.com>

Joscorbe committed:
OAK-9447: cleanup formatting changes

Joscorbe committed:
OAK-9447: cleanup formatting changes

Joscorbe commented:
This PR will likely contain many commits, so plan it to squash it into a single commit when ready.

Joscorbe committed:
OAK-9447: Fix build and removed unwanted changes due to conflicts.

Joscorbe committed:
OAK-9447: Removed some warnings and reverted some unwanted changes

Joscorbe committed:
OAK-9447: Cleanup

Joscorbe committed:
OAK-9447: Some imports ordering and use ping command to check Mongo availability

Joscorbe commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/ClusterInfoTest.java:
```diff
@@ -89,11 +89,11 @@ public void readWriteMode() throws InterruptedException {
 
         op = new UpdateOp(list.get(0).getId(), false);
-        op.set("readWriteMode", "read:nearest, write:fsynced");
+        op.set("readWriteMode", "read:nearest, write:w2");
```
FSYNCED doesn't exist anymore, the correct replacement is W2.

Joscorbe commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/BlobThroughPutTest.java:
```diff
@@ -59,13 +60,11 @@ public class BlobThroughPutTest {
-        bimap.put(WriteConcern.NORMAL,"NORMAL");
 //        bimap.put(WriteConcern.REPLICAS_SAFE,"REPLICAS_SAFE");
-        bimap.put(WriteConcern.SAFE,"SAFE");
+        bimap.put(WriteConcern.ACKNOWLEDGED,"ACKNOWLEDGED");
```
Those WriteConcerns are removed:

- FSYNC_SAFE and JOURNAL_SAFE replaced by JOURNALED
- NORMAL and SAFE replaced by ACKNOWLEDGED

Joscorbe committed:
OAK-9447: reverted some changes in imports and keep functionality.

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStore.java:
```diff
@@ -2361,8 +2364,7 @@ private boolean withClientSession() {
     private boolean secondariesWithinAcceptableLag() {
-        return getClient().getReplicaSetStatus() == null
-                || connection.getStatus().getReplicaSetLagEstimate() < acceptableLagMillis;
+        return connection.getStatus().getReplicaSetLagEstimate() < acceptableLagMillis;
```
getStatus() can't be null.

Joscorbe replied:
getReplicaSetLagEstimate() returns 0 when Mongo is not a replica set.

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java:
```diff
@@ -18,21 +18,20 @@
 import static com.mongodb.client.model.Projections.include;
 import static com.mongodb.client.model.Sorts.ascending;
+
+import static java.util.Collections.emptyList;
```
Not keeping the old imports ordering because they made no sense.

Joscorbe commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStore.java:
```diff
@@ -99,19 +87,30 @@
 import org.slf4j.LoggerFactory;
 
 import com.mongodb.BasicDBObject;
+import com.mongodb.ConnectionString;
```
Not keeping the old imports ordering because they made no sense.

Joscorbe committed:
OAK-9447: revert some formatting changes.

Joscorbe committed:
OAK-9447: avoid formatting changes

Joscorbe committed:
OAK-9447

Joscorbe committed:
OAK-9447: Restore functionality of hasWriteConcern.

Joscorbe committed:
OAK-9447: Some empty lines cleaning and fix ReplicaSetDefaultWriteConcernIT

Joscorbe committed:
OAK-9447: remove empty line

Joscorbe commented:
A few references about the migration:
https://www.mongodb.com/docs/drivers/java/sync/v5.2/reference/upgrade/#breaking-changes
https://www.mongodb.com/docs/drivers/java/sync/v5.2/reference/legacy/#api-changes

Joscorbe committed:
OAK-9447: Bump to version 5.2.1, since 5.2.0 requires an OSGi annotation import.

Joscorbe committed:
OAK-9447: Fix import-packages.

nfsantos reviewed (approved):
The new Java driver seems to work correctly with indexing.

Joscorbe commented:
Thanks @nfsantos!

rishabhdaim commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/DocumentStoreIndexerBase.java:
```diff
@@ -502,7 +502,7 @@ private void configureEstimators(IndexingProgressReporter progressReporter) {
         MongoConnection mongoConnection = indexHelper.getService(MongoConnection.class);
         if (mongoConnection != null) {
-            return mongoConnection.getDatabase().getCollection("nodes").count();
+            return mongoConnection.getDatabase().getCollection("nodes").countDocuments();
```
please use https://mongodb.github.io/mongo-java-driver/5.5/apidocs/driver-sync/com/mongodb/client/MongoCollection.html#estimatedDocumentCount(). `countDocuments` is pretty slow [1].

[1] https://mongodb.github.io/mongo-java-driver/5.5/apidocs/driver-sync/com/mongodb/client/MongoCollection.html#countDocuments()

Joscorbe replied:
You are right, applied, thanks!

rishabhdaim commented on oak-store-document/pom.xml:
```diff
@@ -43,8 +43,8 @@
-              com.mongodb*;version="[3.8, 4)";resolution:=optional,
-              org.bson*;version="[3.8, 4)";resolution:=optional,
+              com.mongodb*;version="[5.2, 5.3)";resolution:=optional,
+              org.bson*;version="[5.2, 5.3)";resolution:=optional,
```
we already have version `5.5.0` released for mongo-driver-sync [1]

[1] https://mvnrepository.com/artifact/org.mongodb/mongodb-driver-sync

Joscorbe replied:
Yes, but we should test it before allowing it. We can do it in a new PR after this is merged, to avoid further delays on this PR.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:
```diff
@@ -340,7 +342,11 @@ private void registerNodeStore() throws IOException {
-            MongoClientURI mongoURI = new MongoClientURI(uri);
+            ConnectionString mongoURI = new ConnectionString(uri);
+            MongoClientSettings settings = MongoClientSettings.builder()
+                    .applyConnectionString(mongoURI)
```
variable `mongoURI` can be inlined.
```suggestion
                    .applyConnectionString(new ConnectionString(uri))
```

Joscorbe replied:
Done

Joscorbe replied:
I had to revert this because the `mongoURI` is [used a few lines below](https://github.com/apache/jackrabbit-oak/pull/2226/files/974372ee2b6d26cfb898864d58eaa45b0a352a7b#diff-0e315af2cc5cc1327210891e5ba43454d739fd8bcdcfefb016d985f071a35cb2R359) in a info log.

rishabhdaim commented on oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java:
```diff
@@ -179,11 +198,8 @@ public static String toString(MongoClientOptions opts) {
-        WriteConcern wc = new MongoClientURI(requireNonNull(uri), builder)
-                .getOptions().getWriteConcern();
-        return !WC_UNKNOWN.equals(wc);
+        ConnectionString connectionString = new ConnectionString(requireNonNull(uri));
```
why do we need need to change this behavior here, this might cause an regression.

Joscorbe replied:
The behavior is the same: both implementations check whether the URI explicitly sets a `writeConcern`. The previous version used a workaround with `WC_UNKNOWN` to detect this, but the new API exposes this check directly via `getWriteConcern()`.

Since now there is a native method in the driver that does exactly what we need, I don't think we need the workaround anymore.

There are tests that checks this and I introduced a few more test cases in [MongoConnectionTest.java](https://github.com/apache/jackrabbit-oak/pull/2226/files/974372ee2b6d26cfb898864d58eaa45b0a352a7b#diff-b17e7e53cb9e04dbff6c53c1bb9b53ae708714314c21b2daf515acf1ee7b5e97) to ensure no regressions are introduced.

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreServiceTest.java:
```diff
@@ -155,18 +153,6 @@ public void setUpdateLimit() throws Exception {
     }
 
-    @Test
-    public void keepAlive() throws Exception {
```
`keepAlive` config is only marked as deprecated not removed. I would suggest to keep these test till we remove that.

Joscorbe replied:
This test checks for `getMongoClientOptions().isSocketKeepAlive()`, which doesn't exist at all in modern driver, so this test would have no sense at all.

I marked `keepAlive` and all related properties as deprecated just to help to detect any uses in any application using Oak as dependency, to avoid breaking changes, but all those options should be completely removed.

rishabhdaim commented on oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoConnectionTest.java:
```diff
@@ -85,26 +95,6 @@ public void sufficientReadConcern() throws Exception {
     }
 
-    @Test
-    public void socketKeepAlive() throws Exception {
```
Again, I would keep this alive till we remove the config.

Joscorbe replied:
As before, Mongo doesn't have any related `isSocketKeepAlive` method anymore. The only way to keep this test is to make it a NOOP test, which doesn't make much sense to me.

Joscorbe replied:
In [Mongo Java Driver 3.6](https://mongodb.github.io/mongo-java-driver/3.6/javadoc/?com/mongodb/MongoClientOptions.html) it was already deprecated, and in 4.0 removed.

Joscorbe committed:
OAK-9447: Restored previous annotations, as they are not always enforced.

Joscorbe committed:
OAK-9447: Applied PR suggestions.

Joscorbe committed:
Merge branch 'trunk' into OAK-9447-joscorbe

Joscorbe committed:
OAK-9447: Add back mongoURI

Joscorbe committed:
Merge remote-tracking branch 'origin/trunk' into OAK-9447-joscorbe

# Conflicts:
#	oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java

Joscorbe committed:
Merge branch 'OAK-9447-joscorbe' of https://github.com/apache/jackrabbit-oak into OAK-9447-joscorbe

Joscorbe committed:
Merge branch 'trunk' into OAK-9447-joscorbe

Joscorbe committed:
OAK-9447: Update MongoFullGcNodeBinSumBsonSizeTest to new Mongo Driver.

Joscorbe committed:
OAK-9447: Solve casting issue in MongoFullGcNodeBinSumBsonSizeTest.

Joscorbe committed:
OAK-9447: Updated Number casting on numeric values returned by Mongo.

This is something that changed in Mongo Java Driver 5.2: https://www.mongodb.com/docs/drivers/java/sync/current/reference/upgrade/#version-5.0-breaking-changes

Joscorbe committed:
OAK-9447: Fix type in MongoFullGcNodeBinSumBsonSizeTest.
